### PR TITLE
Fix/vote for delegate

### PR DESCRIPTION
**Pull request by martiliones**
### `api.voteForDelegate()`

- fix: validation for public keys and delegate names
- chore: better logging
- chore: updated example

### CODE REVIEW

File: src/api/index.ts [hunk 1]
```diff
@@ -499,14 +509,19 @@ export class AdamantApi extends NodeManager {
...
-      if (isAdmVoteForDelegateName(name)) {
+      if (isAdmVoteForDelegateName(vote)) {
         const response = await this.getDelegate({username: name});
 
         if (!response.success) {
           this.logger.warn(
             `[ADAMANT js-api] Failed to get public key for ${vote}. ${response.errorMessage}.`
           );
-          return badParameter('votes');
+          console.log('couldnt get delegates publickey');
```
Comment (adamant-al) [inline src/api/index.ts]: Forgot to clean up?

File: src/api/index.ts [hunk 2]
```diff
@@ -488,7 +493,12 @@ export class AdamantApi extends NodeManager {
           this.logger.warn(
             `[ADAMANT js-api] Failed to get public key for ${vote}. ${response.errorMessage}.`
           );
-          return badParameter('votes');
+
+          return badParameter(
+            'votes',
+            vote,
+            "couldn't get address' public key"
```
Comment (bludnic) [inline src/api/index.ts]: Is this a typo `address'`?

File: src/api/index.ts [hunk 3]
```diff
@@ -454,10 +454,15 @@ export class AdamantApi extends NodeManager {
...
+   *   [
+   *     '+b3d0c0b99f64d0960324089eb678e90d8bcbb3dd8c73ee748e026f8b9a5b5468',
+   *     '-9ef1f6212ae871716cfa2d04e3dc5339e8fe75f89818be21ee1d75004983e2a8',
+   *     '+system',
+   *     '-U16615166477939602094'
+   *   ]
+   * )
    * ```
    */
   async voteForDelegate(passphrase: string, votes: string[]) {
```
Comment (bludnic) [inline src/api/index.ts:471]: Add `@param passphrase Account's passphrase` to JSDoc.

File: src/api/index.ts [hunk 4]
```diff
@@ -517,8 +532,12 @@ export class AdamantApi extends NodeManager {
         continue;
       }
 
-      if (!isAdmVoteForPublicKey(name)) {
-        return badParameter('votes');
+      if (!isAdmVoteForPublicKey(vote)) {
+        return badParameter(
+          'votes',
+          name,
+          "the vote doesn't look like public key, address or delegate name"
```
Comment (bludnic) [inline src/api/index.ts:558]:
```js
const result = await api.voteForDelegate('passphrase...', [
  'U123456'
])
```

Result:

```js
{
  success: false,
  error: "Wrong 'votes' parameter: 123456. Error: the vote doesn't look like public key, address or delegate namethe vote doesn't look like public key, address or delegate name"
}
```

The message is logged two times.

Screenshot 2024-04-24 3 44 24 PM

**Comment by bludnic** [comment]
```js
const result = await api.voteForDelegate('passphrase...', [
  '+U123456'
])
```

**Actual:**

Votes is `+null`

Screenshot 2024-04-24 3 52 34 PM

**Expected:**

Rejected if error if no public key

**Comment by bludnic** [comment]
Can you add a validation to check that the user is actually a Delegate? @martiliones 
Voting for a regular user may result in a waste of ADM tokens.

**Review by bludnic — CHANGES_REQUESTED** [review]

**Comment by adamant-al** [comment]
> Can you add a validation to check that the user is actually a Delegate? @martiliones Voting for a regular user may result in a waste of ADM tokens.

Do we want one more request?

**Comment by martiliones** [comment]
> Do we want one more request?

Well, it can be done using the single request by retrieving all the delegates, since the amount of them is limited to 101, then find the delegate by address there. It doesn't cost much performance for a node since it does exactly the same under the hood for `/delegates/get`

It would be nice if we had `/delegates/get` endpoint optimized for finding a single delegate with addition of `address` parameter

**Comment by adamant-al** [comment]
> It would be nice if we had `/delegates/get` endpoint optimized for finding a single delegate with addition of `address` parameter

You can create a Trello task for this and refer to the comment.

**Review by bludnic — APPROVED** [review]

**Pull request merged**